src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java

```diff
@@ -986,10 +986,10 @@ private void handleOne(final Callback[] callbacks, final int idx) throws IOExcep
                     try {
                         X509Certificate[] x509Certificates = X500.asX509CertificateArray(sslCallback.getSslSession().getPeerCertificates());
                         verifyEvidence(new X509PeerCertificateChainEvidence(x509Certificates));
```

Comment:
The problem that remains here is if verifyEvidence is successful the state will be NameAssignedState - NameAssignedState does not allow you to transition to a different name so no other mechanisms will be able to authenticate.

Also IMO I think it is wasteful to call this for every authentication attempt as this could also involve the realm making remote calls when verifying the certificate even if certificate verification is not required.